### ISSUE TITLE
ntfs3 support?

### ISSUE THREAD

**Issue by BlueMax** [veracrypt/VeraCrypt#847]
Hallo,
how can i mount an NTFS partition with 'ntfs3' (instead of 'ntfs-3g') via GUI?
Manually unmounting and remounting as ntfs3 doesn't work either (mount -t ntfs3 /dev/mapper/veracrypt2 /tmp/folder). Doing this with 'ntfs-3g' does work though.

'ntfs3' has been introduced in kernel 5.15.

Arch Linux 5.15.2
Veracrypt 1.24.update7

**Comment by RealTehreal** [comment]
Could be something on your end. After mounting the volume, without mounting the fs, typing
`mount -t ntfs3 /dev/mapper/veracrypt1 /where/ever`
actually mounts it for me.

Ubuntu 20.04 LTS
VeraCrypt 1.24-update7

**Comment by BlueMax** [comment]
Yes, looks like i have this problem. Will try to fix this on Windows. Veracrypt cannot fix this with its own 'repair' function. And it nevertheless doesn't seem to mount it automatically with NTFS3.
´´´
[12396.796442] ntfs3: dm-1: volume is dirty and "force" flag is not set!
´´´

**Comment by andresbach** [comment]
Has been any news of VeraCrypt supporting `ntfs3` directly in the GUI?
Tried [this](https://wiki.archlinux.org/title/NTFS#unknown_filesystem_type_'ntfs') workaround unsuccessfully.

Arch Linux 6.2.8
VeraCrypt 1.25.9

**Comment by peteh** [comment]
Pushing this up nearly a year later. 

Currently by default veracrypt is not using the ntfs3 driver even if it's available on the system.

**Comment by Alden20** [comment]
Right now, January 2025, I have the same problem (I'm using CachyOS, a derivative of Arch Linux). From VeraCrypt on Linux you can't mount an encrypted partition (or volume) of type NTFS, it always shows the error "unknown filesystem type 'ntfs'". I have tried various options but from the interface I have not been able to get it, no matter how many times I tell VeraCrypt the file system to use (ntfs3) to mount it, it always returns the same error, it seems that it always tries to mount it with the 'ntfs' system and not with 'ntfs3'.

However, from the command line, I have managed to do it in two steps, separating the decryption and the mounting of the partition or volume (note: it is very possible that these steps, literally, only work for Arch Linux or a derivative. You may have to modify some path depending on whether you use another Linux distro):

1- First I decrypt the partition or volume (assuming that I have already created the mounting directory in home/my_user/Data/ and that the path of the partition or volume that I am going to decrypt is /dev/nvme0n1p4):

veracrypt --filesystem=none --mount-options=nokernelcrypto /dev/nvme0n1p4 /home/my_user/Data/ (At this point the VeraCrypt window will appear asking me for the password to decrypt it).

2- Once the previous command has been executed successfully, I mount the unit indicating that it is of type ntfs3:

sudo mount -t ntfs3 /dev/loop0 /home/my_user/Data/ -o rw,nosuid,nodev,relatime,iocharset=utf8 (VeraCrypt, at least in ArchLinux, generates for each decrypted partition or volume a file in /dev/loop and a number depending on how many units are already decrypted at that time).

Once these steps are done, you can now access the decrypted content of the volume or unit in /home/my_user/Data/.

But, as I said, it doesn't work from the interface, it is always trying to mount the decrypted partition or volume with ntfs.

I have tried the following options and none of them have worked:

- Indicate in the interface, in the Mount Options tab of the configuration, that the mount option is: -t ntfs3. It doesn't work.

- I tried to put these lines in the configuration XML file, but it doesn't work either:

<config key="DefaultMountOptions">-t ntfs3</config>
<config key="FilesystemOptions">-t ntfs3 -o rw,nosuid,nodev,relatime,iocharset=utf8</config>
<config key="MountOptions">-t ntfs3</config>

- I tried to mount it with the following option, but it doesn't work either:

veracrypt --filesystem=ntfs3 /dev/nvme0n1p4 /home/my_user/Data/

As I said, the only option that worked for me was the one I indicated above, doing it in two steps from the Linux console, but there's no way to get it from the interface. I can decrypt it from the interface by activating the No Mount option in the window that asks for the password. But from there I have to mount it manually from the console.

**Comment by Jertzukka** [comment]
Whenever you call `mount` with type parameter, the program will always (unless --internal-only is used) use the helper function called `mount.type` so `mount.ntfs` in this case. You can run `mount.ntfs --version` to see what your current default is, and you can run `which mount.ntfs` then run for example `ls -la /usr/bin/mount.ntfs` and it should show where this program is symlinked to if anywhere. If you're using ntfs-3g this is probably symlinked to `/usr/bin/ntfs-3g`, otherwise probably nowhere. By choosing what `mount.ntfs` points to, you should get VeraCrypt to use the preferred version of ntfs driver you'd like to use.

If you'd like to use different mount program than what your mount defaults to, I don't know if that is possible. FreeBSD provides `mountprog` argument for specifying custom helper for this, but I don't think such exists for Linux version.

**Comment by anodynos** [comment]
This should be trivial to fix, at least for the CLI - it should only require one more `ntfs3` option here **https://github.com/veracrypt/VeraCrypt/blob/97a715e8d748d3ad32afa59921d569a127da97a5/src/Main/CommandLineInterface.cpp#L342** as VeraCrypt says `--filesystem=TYPE Filesystem type to mount. The TYPE argument is passed to mount(8) command with option -t.`. Right now this is not recognised, hence we get `Error: Unknown option: ntfs3`

The UI doesn't have the option for `--filesystem` (I guess it defaults to `auto`) - it could be a nice addition!

I used @Alden20 suggestion (BTW, on Kubuntu 22.04, works fine)! cause when I was running a VMware VM from a mounted drive with the old fuse-based `ntfs-3g`, it was so sluggishly slow that it was unusable (NOTE: with or without VeraCrypt, i.e from a normal NTFS  drive mounted as ntfs-3g fuse).  Mounting with the new kernel based `ntfs3` on a VeraCrypt volume (even with the old kernel 5.15) it made the VM usable again. 

The shell script way though is brittle, cause it's not always `/dev/loop0`, in my case it's `/dev/loop29` for some reason. 

Also ntfs3 is not as reliable as the the old ntfs-3g from what I can gather... needs caution!

**Comment by idrassi** [comment]
Thank you everyone for the reports, testing notes, and workarounds shared here over the years.

This has been addressed for Linux in VeraCrypt 1.26.28.

The implementation adds explicit in-kernel NTFS driver selection for mounts:

- `--filesystem=ntfs3` pins the Linux kernel `ntfs3` driver and bypasses mount helpers such as `mount.ntfs3`.
- `--filesystem=kernel-ntfs` selects an available in-kernel NTFS driver for an NTFS volume.
- `-m kernelntfs` enables the same behavior for the current mount when NTFS is detected and no explicit filesystem type was supplied.
- GUI users can enable `Settings -> Preferences -> Mount Options -> Mount NTFS volumes with an in-kernel Linux driver`.

Examples:

```bash
veracrypt -t --filesystem=ntfs3 volume.hc /media/veracrypt1
```
or:

```bash
veracrypt -t --filesystem=kernel-ntfs volume.hc /media/veracrypt1
```

This is an opt-in path and doesn't change the default NTFS backend unless one of the options above, or the GUI preference, is used. It also doesn't override normal kernel NTFS behavior for dirty or hibernated NTFS volumes.

I will close this issue as fixed in 1.26.28. If anyone still sees failures with [1.26.28 beta4](https://github.com/veracrypt/VeraCrypt/releases/tag/VeraCrypt_1.26.28_Beta4)  or later, please open a new issue with the distribution, kernel version, VeraCrypt version, the exact command or GUI option used, and the full error output, including relevant dmesg output if available.
